File: move_gen.py
import bitboard

# Push a pawn forward by a single square based on player turn
def single_push_pawn(player):
    occupied = bitboard.occupied_squares()

    if player.turn:
        return bitboard.north(bitboard.WHITE_PAWNS) & ~occupied & ~bitboard.RANK_8
    else:
        return bitboard.south(bitboard.BLACK_PAWNS) & ~occupied & ~bitboard.RANK_1

# Push a pawn forward by a two squares based on player turn
def double_push_pawn(player):
    occupied = bitboard.occupied_squares()

    if player.turn:
        single_push = bitboard.north(bitboard.WHITE_PAWNS) & ~occupied
        return (single_push << 8) & ~occupied & bitboard.RANK_4

    else:
        single_push = bitboard.south(bitboard.BLACK_PAWNS) & ~occupied
        return (single_push >> 8) & ~occupied & bitboard.RANK_5
    
# Generate an attack table for pawn attacks
def generate_pawn_attacks(player, bb):
    attacks = [0, 0]
        
    # Check white pawn attacks
    if player == 0:
        attacks[0] = bitboard.north_east(bb) | bitboard.north_west(bb)

    # Check black pawn attacks
    elif player == 1:
        attacks[1] = bitboard.south_west(bb) | bitboard.south_east(bb)

    return attacks

# Initialize attack tables
def init_attack_table_pawn():
    attacks = [[0 for _ in range(64)] for _ in range(2)]

    for sq in range(64):
        square = bitboard.SQUARES[sq]
        attacks[0][sq] = generate_pawn_attacks(0, square)[0]
        attacks[1][sq] = generate_pawn_attacks(1, square)[1]

    return attacks

# Generate an attack table for knight attacks
def generate_knight_attacks():
    attacks = [0] * 64
    
    for sq in range(64):
        square = bitboard.SQUARES[sq]
        attacks[sq] = bitboard.north_ne(square) | bitboard.north_nw(square) | \
                      bitboard.north_ee(square) | bitboard.north_ww(square) | \
                      bitboard.south_se(square) | bitboard.south_sw(square) | \
                      bitboard.south_ee(square) | bitboard.south_ww(square)

    return attacks

# Generate an attack table for king attacks
def generate_king_attacks():
    attacks = [0] * 64

    for sq in range(64):
        square = bitboard.SQUARES[sq]
        attacks[sq] = bitboard.north(square) | bitboard.south(square) | \
                      bitboard.east(square) | bitboard.west(square) | \
                      bitboard.north_east(square) | bitboard.north_west(square) | \
                      bitboard.south_east(square) | bitboard.south_west(square)
        
    return attacks

# Generate bishop masks
def generate_bishop_mask(square):
    mask = 0
    file_mask = square % 8
    rank_mask = square // 8

    for i in range(1, 7):
        if (rank_mask + i <= 6) & (file_mask + i <= 6):
            mask |= bitboard.SQUARES[(rank_mask + i) * 8 + (file_mask + i)]
        if rank_mask - i >= 1 and file_mask - i >= 1:
            mask |= bitboard.SQUARES[(rank_mask - i) * 8 + (file_mask - i)]
        if rank_mask + i <= 6 and file_mask - i >= 1:
            mask |= bitboard.SQUARES[(rank_mask + i) * 8 + (file_mask - i)]
        if rank_mask - i >= 1 and file_mask + i <= 6: 
            mask |= bitboard.SQUARES[(rank_mask - i) * 8 + (file_mask + i)]

    return mask

# Initialize bishop mask table
def init_bishop_mask():
    mask = [0] * 64

    for sq in range(64):
        mask[sq] = generate_bishop_mask(sq)

    return mask

# Generate rook masks
def generate_rook_mask(square):
    mask = 0
    file_mask = square % 8
    rank_mask = square // 8

    for i in range(1, 7):
        if rank_mask + i <= 6:
            mask |= bitboard.SQUARES[(rank_mask + i) * 8 + file_mask]
        if rank_mask - i >= 1:
            mask |= bitboard.SQUARES[(rank_mask - i) * 8 + file_mask]
        if file_mask - i >= 1:
            mask |= bitboard.SQUARES[rank_mask * 8 + (file_mask - i)]
        if file_mask + i <= 6: 
            mask |= bitboard.SQUARES[rank_mask * 8 + (file_mask + i)]

    return mask

# Initialize rook mask table
def init_rook_mask():
    mask = [0] * 64

    for sq in range(64):
        mask[sq] = generate_rook_mask(sq)

    return mask

File: move_gen.cpp
#include "bitboard.hpp"
#include "move_gen.hpp"
#include <vector>
#include <cstdlib>
#include <array>

// Define tables
std::vector<std::vector<uint64_t>> pawnAttackTable(2, std::vector<uint64_t>(64, 0));
std::array<uint64_t, 64> knightAttackTable{};
std::array<uint64_t, 64> kingAttackTable{};
std::array<uint64_t, 64> bishopMaskTable{};
std::array<uint64_t, 64> rookMaskTable{};
std::array<uint64_t, 64> queenMaskTable{};

// Bishop and rook occupancy tables
const std::vector<int> bishopOccupancy = {
    6, 6, 6, 6, 6, 6, 6, 6,
    6, 5, 5, 5, 5, 5, 5, 6,
    6, 5, 7, 7, 7, 7, 5, 6,
    6, 5, 7, 9, 9, 7, 5, 6,
    6, 5, 7, 9, 9, 7, 5, 6,
    6, 5, 7, 7, 7, 7, 5, 6,
    6, 5, 5, 5, 5, 5, 5, 6,
    6, 6, 6, 6, 6, 6, 6, 6
};

const std::vector<int> rookOccupancy = {
    12, 11, 11, 11, 11, 11, 11, 12,
    11, 10, 10, 10, 10, 10, 10, 11,
    11, 10, 10, 10, 10, 10, 10, 11,
    11, 10, 10, 10, 10, 10, 10, 11,
    11, 10, 10, 10, 10, 10, 10, 11,
    11, 10, 10, 10, 10, 10, 10, 11,
    11, 10, 10, 10, 10, 10, 10, 11,
    12, 11, 11, 11, 11, 11, 11, 12
};

// Push a pawn forward by one square
uint64_t singlePushPawn(bool isWhite, uint64_t occupied) {
    if (isWhite) {
        return north(WHITE_PAWNS) & ~occupied & ~RANK_8;
    } else {
        return south(BLACK_PAWNS) & ~occupied & ~RANK_1;
    }
}

// Push a pawn forward by two squares
uint64_t doublePushPawn(bool isWhite, uint64_t occupied) {
    if (isWhite) {
        uint64_t singlePush = north(WHITE_PAWNS) & ~occupied;
        return (singlePush << 8) & ~occupied & RANK_4;
    } else {
        uint64_t singlePush = south(BLACK_PAWNS) & ~occupied;
        return (singlePush >> 8) & ~occupied & RANK_5;
    }
}

// Generate pawn attacks
uint64_t generatePawnAttacks(bool isWhite, uint64_t bb) {
    if (isWhite) {
        return (northEast(bb) | northWest(bb)) & ~RANK_8;
    } else {
        return (southWest(bb) | southEast(bb)) & ~RANK_1;
    }
}

std::vector<std::vector<uint64_t>> initPawnAttackTable() {    
    for (int square = 0; square < 64; ++square) {
        uint64_t bb = SQUARES[square];
        pawnAttackTable[1][square] = generatePawnAttacks(true, bb);
        pawnAttackTable[0][square] = generatePawnAttacks(false, bb);
    }
    return pawnAttackTable;
}

// Generate knight attacks
std::array<uint64_t, 64> generateKnightAttacks() {
    for (int sq = 0; sq < 64; ++sq) {
        uint64_t square = SQUARES[sq];
        knightAttackTable[sq] = (northNE(square) | northNW(square) | northEE(square) | northWW(square) |
                       southSE(square) | southSW(square) | southEE(square) | southWW(square));
    }
    return knightAttackTable;
}

// Generate king attacks
std::array<uint64_t, 64> generateKingAttacks() {    
    for (int sq = 0; sq < 64; ++sq) {
        uint64_t square = SQUARES[sq];
        kingAttackTable[sq] = north(square) | south(square) | east(square) | west(square) |
                      northEast(square) | northWest(square) | southEast(square) | southWest(square);
    }
    return kingAttackTable;
}

// Generate bishop mask
uint64_t generateBishopMask(int square, uint64_t blockers) {
    uint64_t mask = 0;
    std::vector<uint64_t (*)(uint64_t)> directions = { northEast, northWest, southEast, southWest };
    
    for (auto direction : directions) {
        uint64_t bb = SQUARES[square];
        while (true) {
            bb = direction(bb);
            if (bb == 0) break;
            mask |= bb;
            if (bb & blockers) break;
        }
    }
    return mask;
}

// Generate rook mask
uint64_t generateRookMask(int square, uint64_t blockers) {
    uint64_t mask = 0;
    std::vector<uint64_t (*)(uint64_t)> directions = { north, south, east, west };
    
    for (auto direction : directions) {
        uint64_t bb = SQUARES[square];
        while (true) {
            bb = direction(bb);
            if (bb == 0) break;
            mask |= bb;
            if (bb & blockers) break;
        }
    }
    return mask;
}

// Initialize tables
void initMaskTables() {
    for (int square = 0; square < 64; ++square) {
        bishopMaskTable[square] = generateBishopMask(square, 0);
        rookMaskTable[square] = generateRookMask(square, 0);
        queenMaskTable[square] = bishopMaskTable[square] | rookMaskTable[square];
    }
}

void initAttackTables() {
    initPawnAttackTable();
    generateKnightAttacks();
    generateKingAttacks();
}

// Set occupancy bits
uint64_t setOccupancy(int index, int bits, uint64_t attacks) {
    uint64_t occupancy = 0;

    for (int i = 0; i < bits; ++i) {
        int square = getLSB(attacks);
        attacks = clearPiece(attacks, square);

        if (index & (1 << i)) {
            occupancy |= 1ULL << square;
        }
    }

    return occupancy;
}

// Pseudo-random number generator (32-bit)
uint32_t xorshift32RNG(uint32_t& state) {
    state ^= state << 13;
    state ^= state >> 7;
    state ^= state << 17;
    return state & 0xFFFFFFFF;
}

// Pseudo-random number generator (64-bit)
uint64_t xorshift64RNG(uint32_t& state) {
    uint64_t num1 = xorshift32RNG(state) & 0xFFFF;
    uint64_t num2 = xorshift32RNG(state) & 0xFFFF;
    uint64_t num3 = xorshift32RNG(state) & 0xFFFF;
    uint64_t num4 = xorshift32RNG(state) & 0xFFFF;

    return (num1 << 48) | (num2 << 32) | (num3 << 16) | num4;
}

// Generate magic number
uint64_t getMagicNumber(uint32_t& state) {
    return xorshift64RNG(state) & xorshift64RNG(state) & xorshift64RNG(state);
}